homeassistant/components/tuya/config_flow.py

```diff
@@ -106,3 +146,261 @@ async def async_step_user(self, user_input=None):
+        """Get the list of Tuya device to filtered by types."""
+        config_list = {}
+        types_filter = set(types)
+        tuya = self.hass.data[DOMAIN][TUYA_DATA]
+        devices_list = tuya.get_all_devices()
+        for device in devices_list:
+            dev_type = device.device_type()
+            exclude = (
+                dev_type in types_filter
+                if exclude_mode
+                else dev_type not in types_filter
+            )
+            if exclude:
+                continue
+            dev_id = device.object_id()
+            if type_prefix:
+                dev_id = f"{dev_type}-{dev_id}"
+            config_list[dev_id] = f"{device.name()} ({dev_type})"
+
+        return config_list
+
+    def _get_device(self, dev_id):
+        """Get specific device from tuya library."""
+        tuya = self.hass.data[DOMAIN][TUYA_DATA]
+        return tuya.get_device_by_id(dev_id)
+
+    def _save_config(self, data):
+        """Save the updated options."""
+        curr_conf = self.config_entry.options.copy()
+        curr_conf.update(data)
+        curr_conf.update(self._conf_devs_option)
+
+        return self.async_create_entry(title="", data=curr_conf)
+
+    async def _async_device_form(self, devs_id):
+        """Return configuration form for devices."""
+        conf_devs_id = []
+        for count, dev_id in enumerate(devs_id):
+            device_info = dev_id.split("-")
+            if count == 0:
+                device_type = device_info[0]
+                device_id = device_info[1]
+            elif device_type != device_info[0]:
+                self._form_error = ERROR_DEV_MULTI_TYPE
+                return await self.async_step_init()
+            conf_devs_id.append(device_info[1])
+
+        device = self._get_device(device_id)
+        if not device:
+            self._form_error = ERROR_DEV_NOT_FOUND
+            return await self.async_step_init()
+
+        curr_conf = self._conf_devs_option.get(
+            device_id, self.config_entry.options.get(device_id, {})
+        )
+
+        config_schema = await self._get_device_schema(device_type, curr_conf, device)
+        if not config_schema:
+            self._form_error = ERROR_DEV_NOT_CONFIG
+            return await self.async_step_init()
+
+        self._conf_devs_id = conf_devs_id
+        device_name = (
+            "(multiple devices selected)" if len(conf_devs_id) > 1 else device.name()
+        )
+
+        return self.async_show_form(
+            step_id="device",
+            data_schema=config_schema,
+            description_placeholders={
+                "device_type": device_type,
+                "device_name": device_name,
+            },
+        )
+
+    async def async_step_init(self, user_input=None):
+        """Handle options flow."""
+        if user_input is not None:
+            dev_ids = user_input.get(CONF_LIST_DEVICES)
+            if dev_ids:
+                return await self._async_device_form(dev_ids)
```

Comment:
We aren't allowed to show the form for another step.  `_async_device_form` currently displays the form for the `device` step.  

Please move to the `device` step by calling `await self.async_step_device(...)` before showing the form.

Reply:
Done